src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs

```diff
@@ -218,93 +221,202 @@ public static unsafe int IndexOf(ref char searchSpace, char value, int length)
 
-                if (Vector.IsHardwareAccelerated && length >= Vector<ushort>.Count * 2)
+            if (Avx2.IsSupported || Sse2.IsSupported)
+            {
+                // Avx2 branch also operates on Sse2 sizes, so check is combined.
+                if (length >= Vector128<byte>.Count * 2)
                 {
-                    // Figure out how many characters to read sequentially until we are vector aligned
-                    // This is equivalent to:
-                    //         unaligned = ((int)pCh % Unsafe.SizeOf<Vector<ushort>>()) / elementsPerByte
-                    //         length = (Vector<ushort>.Count - unaligned) % Vector<ushort>.Count
-                    const int elementsPerByte = sizeof(ushort) / sizeof(byte);
-                    int unaligned = ((int)pCh & (Unsafe.SizeOf<Vector<ushort>>() - 1)) / elementsPerByte;
-                    length = (Vector<ushort>.Count - unaligned) & (Vector<ushort>.Count - 1);
+                    lengthToExamine = UnalignedCountVector128(ref searchSpace);
                 }
-
-            SequentialScan:
-                while (length >= 4)
+            }
+            else if (Vector.IsHardwareAccelerated)
+            {
+                if (length >= Vector<ushort>.Count * 2)
                 {
-                    length -= 4;
+                    lengthToExamine = UnalignedCountVector(ref searchSpace);
+                }
+            }
 
-                    if (pCh[0] == value)
-                        goto Found;
-                    if (pCh[1] == value)
-                        goto Found1;
-                    if (pCh[2] == value)
-                        goto Found2;
-                    if (pCh[3] == value)
-                        goto Found3;
+        SequentialScan:
+            while (lengthToExamine >= 4)
+            {
+                ref char current = ref Add(ref searchSpace, offset);
+                lengthToExamine -= 4;
+
+                if (value == current)
+                    goto Found;
+                if (value == Add(ref current, 1))
+                    goto Found1;
+                if (value == Add(ref current, 2))
+                    goto Found2;
+                if (value == Add(ref current, 3))
+                    goto Found3;
+
+                offset += 4;
+            }
 
-                    pCh += 4;
-                }
+            while (lengthToExamine > 0)
+            {
+                lengthToExamine -= 1;
 
-                while (length > 0)
+                if (value == Add(ref searchSpace, offset))
+                    goto Found;
+
+                offset += 1;
+            }
+
+            // We get past SequentialScan only if IsHardwareAccelerated or intrinsic .IsSupported is true. However, we still have the redundant check to allow
+            // the JIT to see that the code is unreachable and eliminate it when the platform does not have hardware accelerated.
+            if (Avx2.IsSupported)
```

Comment:
Is using AVX2 always more beneficial than using SSE2? I've frequently seen cases where SSE2 is faster for small counts and AVX2 starts taking over at some slightly higher limit.

Reply:
This path uses both Avx2 and Sse2

Reply:
Right. But I was asking if there was a cutoff point at which AVX2 starts being more efficient than SSE2; or if it is just always more beneficial.

For example, I've frequently seen the case where 128-bit code is faster for 16-64 byte inputs and 256-bit code only starts winning out at the 64 or 128-byte mark (sometimes this is higher).

Reply:
Didn't measure any disadvantage
```
 ------------------- |--------- |-----------:|
      String_IndexOf |       16 |   4.697 ns |
      String_IndexOf |       31 |   5.226 ns |
      String_IndexOf |       32 |   5.224 ns |
      String_IndexOf |       63 |   6.991 ns |
      String_IndexOf |       64 |   6.990 ns |
      String_IndexOf |      127 |   9.050 ns |
      String_IndexOf |      128 |   9.228 ns |
```